src/workerd/api/actor.h

```diff
@@ -13,27 +13,95 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
+public:
+  DurableObjectRpcStub(
+      IoOwn<OutgoingFactory> outgoingFactory,
+      RequiresHostAndProtocol requiresHost,
+      IsInHouse inHouse,
+      CompatibilityFlags::Reader flags)
+    : Fetcher(kj::mv(outgoingFactory), requiresHost, (inHouse == YES)), flags(flags) {}
+
+  bool isRegisteredName(kj::StringPtr name) {
+    // We don't need to check `fetch` or `constructor` because `fetch` is known at compile time and
```

Comment:
This comment isn't really accurate with respect to `constructor`, however, it the code is correct.

`constructor` is only a keyword is certain contexts. It is perfectly valid to define a JS object with a property named `constructor` and then access it as a property:

```
Welcome to Node.js v18.13.0.
Type ".help" for more information.
> foo = {constructor: 123}
{ constructor: 123 }
> foo.constructor
123
> 
```

But for the same reason, it wouldn't make sense to block `consturctor` here -- if someone puts a property called `constructor` on their object server-side, we might as well let them call it.

Note that when you declare a class with a constructor:

```
class Foo {
  consturctor() { ... }
}
```

This does NOT declare a method called `constructor`. Within the `class` syntax, the word `constructor` is treated specially.

Also note that the `Object` _prototype_ declares a property called `constructor`, and as a result, for any object O, if O does not itself have a property called `constructor`, then `O.constructor` returns the prototype property. We do _not_ want to make this property available over RPC. However, the same is true of _all_ `Object` prototype properties -- we don't want any of these to be remotely accessible. Making a blocklist of `Object` prototype properties isn't the right way to do that -- we need to make sure, on the server side, that when we look up the method, we do not consider properties of the `Object` prototype.

Reply:
FWIW, it's not so much that this is *blocking* `constructor`, it's that the name intercept is configured such that it is only invoked on properties that are not already declared/known on the object. Since the `Object` prototype has a property that is explicitly named `constructor`, I do not believe the named intercept mechanism will ever be invoked for that property. However, this is a behavior we have complete control over. It's based on the flags that we pass in when configuring the intercept callbacks.

Reply:
Removed the comment since we'll be checking the names on the server side.